File: src/main/java/com/kingdom/service/LoggedInUsers.java
package com.kingdom.service;

import com.kingdom.model.User;

import java.util.*;

public class LoggedInUsers {
    private static Map<Integer, User> users = new HashMap<Integer, User>();

    private static LoggedInUsers ourInstance = new LoggedInUsers();

    public static LoggedInUsers getInstance() {
        return ourInstance;
    }

    public void userLoggedIn(User user) {
        updateUser(user, true, true);
    }

    public void userLoggedOut(User user) {
        users.remove(user.getUserId());
    }

    public void gameReset(int userId) {
        User user = users.get(userId);
        if (user != null) {
            user.setGameId(0);
            user.setLastActivity(new Date());
        }
    }

    public List<User> getUsers() {
        return new ArrayList<User>(users.values());
    }

    public User getUser(int userId) {
        return users.get(userId);
    }

    public void updateUser(User user) {
        updateUser(user, true, false);
    }

    public void refreshLobby(User user) {
        updateUser(user, false, false);
    }

    public void updateUserStatus(User user) {
        updateUser(user, true, true);
    }

    private void updateUser(User user, boolean refreshLastActivity, boolean updateStatus) {
        User loggedInUser = users.get(user.getUserId());
        if (loggedInUser == null) {
            loggedInUser = user;
        }
        loggedInUser.setGameId(user.getGameId());
        if (updateStatus) {
            loggedInUser.setStatus(user.getStatus());
        }
        if (refreshLastActivity) {
            loggedInUser.setLastActivity(new Date());
        }
        loggedInUser.setLastRefresh(new Date());
        users.put(loggedInUser.getUserId(), loggedInUser);
    }

    public void refreshLobbyPlayers() {
        for (User user : users.values()) {
            user.getRefreshLobby().setRefreshPlayers(true);
        }
    }

    public void refreshLobbyGameRooms() {
        for (User user : users.values()) {
            user.getRefreshLobby().setRefreshGameRooms(true);
        }
    }

    public void refreshLobbyChat() {
        for (User user : users.values()) {
            user.getRefreshLobby().setRefreshChat(true);
        }
    }
}

File: src/main/kotlin/com/kingdom/service/LoggedInUsers.kt
package com.kingdom.service

import com.kingdom.model.User
import com.kingdom.util.removeSpaces
import java.util.*

object LoggedInUsers {

    private val users = HashMap<String, User>()

    fun userLoggedIn(user: User) {
        updateUser(user, true, true)
    }

    fun userLoggedOut(user: User) {
        users.remove(user.userId)
    }

    fun gameReset(userId: String) {
        val user = users[userId]
        if (user != null) {
            user.gameId = null
            user.lastActivity = Date()
        }
    }

    fun getUsers(): List<User> {
        return ArrayList(users.values)
    }

    fun getUser(userId: String): User? {
        return users[userId]
    }

    fun usernameBeingUsed(username: String): Boolean {
        return users.values.any { it.username.removeSpaces().toLowerCase() == username.removeSpaces().toLowerCase() }
    }

    fun getUserByUsername(username: String): User? {
        return users.values.firstOrNull { it.username == username }
    }

    fun updateUser(user: User) {
        updateUser(user, true, false)
    }

    fun refreshLobby(user: User) {
        updateUser(user, false, false)
    }

    fun updateUserStatus(user: User) {
        updateUser(user, true, true)
    }

    private fun updateUser(user: User, refreshLastActivity: Boolean, updateStatus: Boolean) {
        var loggedInUser: User? = users[user.userId]
        if (loggedInUser == null) {
            loggedInUser = user
        }
        loggedInUser.gameId = user.gameId
        if (updateStatus) {
            loggedInUser.status = user.status
        }
        if (refreshLastActivity) {
            loggedInUser.lastActivity = Date()
        }
        loggedInUser.lastRefresh = Date()
        users[loggedInUser.userId] = loggedInUser
    }

    fun refreshLobbyPlayers() {
        for (user in users.values) {
            user.refreshLobby.isRefreshPlayers = true
        }
    }

    fun refreshLobbyGameRooms() {
        for (user in users.values) {
            user.refreshLobby.isRefreshGameRooms = true
        }
    }

    fun refreshLobbyChat() {
        for (user in users.values) {
            user.refreshLobby.isRefreshChat = true
        }
    }
}
